[Part 1] feat: updates text field maxLines to double

gaganyadav80 opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Part of #35064 

This PR aims to allow using double instead of int for the max/min line properties of Text Field which will include the following properties:

- maxLines
- minLines
- hintMaxLines
- errorMaxLines

I am aiming to raise separate PR these property changes and this is a part of them. I will keep updating description to include  links to other PR as well.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

gaganyadav80 committed:
feat: updates text field maxLines to double

gaganyadav80 commented on packages/flutter/lib/src/material/selectable_text.dart:
```diff
@@ -774,7 +774,7 @@ class _SelectableTextState extends State<SelectableText>
         forceLine: false,
         minLines: widget.minLines,
-        maxLines: widget.maxLines ?? defaultTextStyle.maxLines,
+        maxLines: (widget.maxLines ?? defaultTextStyle.maxLines)?.toDouble(),
```
I have not updated the maxLines property for `Text/SelectableText/DefaultTextStyle` hence we need to convert them to double before passing to the `EditableText`

gaganyadav80 commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -8560,15 +8560,15 @@ void main() {
               width: 400,
               child: EditableText(
-                minLines: lines,
+                minLines: lines.floor(),
```
Same, will update in other PR

gaganyadav80 commented:
Hello @loic-sharma @dkwingsmt , I wanted to ask if it is alright that I also update the **other** properties (maxLines, minLines, helperMaxLines, and hintMaxLines) to double in this PR.

Because I have already updated `TextField.maxLines` and `Text/SelectableText/DefaultTextStyle.maxLines` in this PR. The number of files will increase but most of these are one line changes. 

I am fine with raising separate PRs if that is preferred. Please let me know.

dkwingsmt commented:
I'd say yes, they should be updated in the same way. However I recommend we only work on one field in this PR to settle the details, and mirror the change to others after this PR.

gaganyadav80 commented:
@dkwingsmt I updated the `Text`/`SelectableText`/`DefaultTextStyle` as @loic-sharma mentioned, along with the changes in the engine for `ui.ParagraphStyle`. 

The update is comprising 24 files change. Should I push in this PR or in a separate PR?

PS: I appreciate your patience if any of my questions seem overly basic. Please let me know if I should adjust my approach.

dkwingsmt commented:
Another question: I see like 80% of changes in this PR are changing int literals to doubles. Are they needed for compiling or just for aesthetic reasons? If it's the latter I would slightly prefer reverting these changes to make the PR cleaner and make subsequent PRs easier for you.

dkwingsmt commented:
Hi!  @gaganyadav80 Are you able to address the comments above? I think it's pretty close to completion after that and I'm looking forward to it!

gaganyadav80 committed:
revert: changes int literals to double

gaganyadav80 committed:
test: update debugFillProperties checks

gaganyadav80 commented:
Hi @dkwingsmt, I have updated the PR to remove the literal changes. But I have not pushed the changes on `Text`/`SelectableText`/`DefaultTextStyle` to update their `maxLines` to double as mentioned by @loic-sharma. Though I have them ready locally.

Let me know if I should push them in this same PR or you would like a new PR :)

dkwingsmt commented:
I think they should be updated in a separate PR. Thanks for your work! Let me take a look now.

gaganyadav80 commented:
Yes, I will push the changes before EOD.

gaganyadav80 committed:
feat: update maxLines to double everywhere

- updated Text, RichText, SelectableText, DefaultTextStyle

- updated helperMaxLines, errorMaxLines and hintMaxLines of InputDecoration

- updated RenderParagraph, TextPainter

- updated ParagraphStyle in flutter engine

gaaclarke commented:
> From the build logs, it looks like I’ll need to pass `maxLines` (now a double) separately through the `ParagraphBuilder`, similar to how `height` and `fontSize` are handled.
> 
> Before I dig deeper into the code, I’d really appreciate your suggestions.

Yes, that's what you need to do.  ParagraphStyle stores an array of int32's that is most of its value encoded.  Max lines is no longer an int32 so it can not longer be stored in there and must be handled like the other double fields, like height.

dkwingsmt commented on engine/src/flutter/lib/web_ui/skwasm/text/paragraph_style.cpp:
```diff
@@ -38,7 +38,7 @@ SKWASM_EXPORT void paragraphStyle_setTextDirection(
 SKWASM_EXPORT void paragraphStyle_setMaxLines(ParagraphStyle* style,
-                                              size_t maxLines) {
+                                              SkScalar maxLines) {
   style->skiaParagraphStyle.setMaxLines(maxLines);
```
It seems that `setMaxLines` is a `size_t` ([link](https://skia.googlesource.com/skia/+/52130e5c3b34/modules/skparagraph/include/ParagraphStyle.h#103)). This call probably truncates the fraction parts. Should we to a `ceil` here instead?

dkwingsmt reviewed (commented):
I propose that we should instead keep engine-level `maxLines` as integers. The engine probably needs to render the full line anyway, and the framework works by clipping the render result. Some, if not all, backend renderers (such as Skia) only supports integral maxLines.

gaganyadav80 committed:
revert: changes to flutter engine

dkwingsmt commented:
Oh, sorry I thought I had replied to you (probably drafted but never sent out).

> To avoid future confusion about why maxLines is an integer in the engine but a double in the framework, would it make sense to add a comment in the framework code explaining this?

It can be useful, if you can think of a good place to explain it (maybe `Text`?). Not strongly necessary though, since it might be intuitive enough to realize that the engine has to draw full text lines but widgets can crop them by partial lines.

dkwingsmt commented:
Yeah, that's unfortunate. Can you try if it builds with `num`?

gaganyadav80 committed:
refactor: change maxLines type from double to num

gaganyadav80 commented:
Hello @dkwingsmt, hope you are doing well. I was just checking in if you got the time to review my last comment?

dkwingsmt commented:
I'm really sorry that your comment slipped through my review queue (I didn't check the "Reviewed" tab).

That's unfortunate indeed. In that case we might have to create a new parameter, for example with the name of `maxDisplayedLines`, that uses `double?`.
* `maxDisplayedLines` should be provided exclusively with `maxLines`, guarded by assertion.
* `get maxLines` should return `ceil(maxDisplayedLines)`.

The biggest downside I can think of is how many classes this will affect.

I thought about hard breaking, but it'll probably go bad considering how widespread these classes have been used for. 

Also cc @loic-sharma and @justinmc.

gaganyadav80 commented:
@dkwingsmt @loic-sharma Thank you for the suggested approach. I will update the PR soon.

dkwingsmt commented:
I'm going to reluctantly close this pull request for now until we're confident enough about this change. I still agree that this is a feature we should support, and the existing choice of `int` parameters is unfortunate. However, given the  complexity (allowing packages to `implement` this class), it is indeed extremely difficult to make such a seemingly small change. We can't afford to break code for such a core feature either. As @loic-sharma pointed out, if you're interested in pushing for this change and would like to draft a design doc that evaluates a proposed solution, we're happy to look at it. If not, that's also okay. I sincerely appreciate the effort you've put into this PR. Thank you very much.
